src/mds/CDir.cc

```diff
@@ -1552,38 +1558,76 @@ void CDir::fetch(MDSContext *c, std::string_view want_dn, bool ignore_authpinnab
   auth_pin(this);
   state_set(CDir::STATE_FETCHING);
 
-  if (mdcache->mds->logger) mdcache->mds->logger->inc(l_mds_dir_fetch);
+  _omap_fetch(nullptr, nullptr);
 
+  if (mdcache->mds->logger)
+    mdcache->mds->logger->inc(l_mds_dir_fetch_complete);
   mdcache->mds->balancer->hit_dir(this, META_POP_FETCH);
-
-  std::set<dentry_key_t> empty;
-  _omap_fetch(NULL, empty);
 }
 
-void CDir::fetch(MDSContext *c, const std::set<dentry_key_t>& keys)
+void CDir::fetch_keys(const std::vector<dentry_key_t>& keys, MDSContext *c)
 {
   dout(10) << "fetch " << keys.size() << " keys on " << *this << dendl;
-
   ceph_assert(is_auth());
   ceph_assert(!is_complete());
 
-  if (!can_auth_pin()) {
-    dout(7) << "fetch keys waiting for authpinnable" << dendl;
-    add_waiter(WAIT_UNFREEZE, c);
+  if (CDir *pdir = get_parent_dir();
+      pdir && pdir->inode->is_stray() && !inode->snaprealm) {
+    fetch(c, true);
     return;
   }
+
+  MDSContext::vec_alloc<mempool::mds_co::pool_allocator> *fallback_waiting = nullptr;
+  std::set<std::string> str_keys;
+  for (auto& key : keys) {
+    ceph_assert(key.snapid == CEPH_NOSNAP);
+    if (waiting_on_dentry.empty())
+      get(PIN_DNWAITER);
+    auto em = waiting_on_dentry.emplace(std::piecewise_construct,
+					std::forward_as_tuple(key.name, key.snapid),
+					std::forward_as_tuple());
+    if (!em.second) {
+      if (!fallback_waiting)
+	fallback_waiting = &em.first->second;
+      continue;
+    }
+
+    if (c) {
+      em.first->second.push_back(c);
+      c = nullptr;
+    }
+
+    string str;
+    key.encode(str);
+    str_keys.emplace(std::move(str));
+  }
+
+  if (str_keys.empty()) {
+    if (c && fallback_waiting) {
+      fallback_waiting->push_back(c);
+      c = nullptr;
+    }
+
+    if (get_version() > 0) {
+      dout(7) << "fetch keys, all are already being fetched" << dendl;
+      ceph_assert(!c);
```

Comment:
When keys.size() == 0，c will be not empty there.

Reply:
Please see https://github.com/ceph/ceph/pull/46063

Reply:
> Please see #46063

Thanks.